File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-7_tooling=beads/rep2/src/queries/stats.ts
import type { Match } from "../types.js";

export interface StatsFilter {
  competition?: string;
  season?: number;
  seasonFrom?: number;
  seasonTo?: number;
}

function filterMatches(matches: Match[], filter: StatsFilter): Match[] {
  const needle = filter.competition?.toLowerCase();
  return matches.filter((m) => {
    if (m.homeGoal === null || m.awayGoal === null) return false;
    if (needle && !String(m.competition).toLowerCase().includes(needle)) return false;
    if (filter.season !== undefined && m.season !== filter.season) return false;
    if (filter.seasonFrom !== undefined && (m.season ?? -Infinity) < filter.seasonFrom) return false;
    if (filter.seasonTo !== undefined && (m.season ?? Infinity) > filter.seasonTo) return false;
    return true;
  });
}

export interface AggregateStats {
  matches: number;
  totalGoals: number;
  averageGoalsPerMatch: number;
  homeWins: number;
  awayWins: number;
  draws: number;
  homeWinRate: number;
  awayWinRate: number;
  drawRate: number;
}

export function aggregateStats(matches: Match[], filter: StatsFilter = {}): AggregateStats {
  const list = filterMatches(matches, filter);
  let totalGoals = 0, homeWins = 0, awayWins = 0, draws = 0;
  for (const m of list) {
    totalGoals += (m.homeGoal ?? 0) + (m.awayGoal ?? 0);
    if ((m.homeGoal ?? 0) > (m.awayGoal ?? 0)) homeWins++;
    else if ((m.homeGoal ?? 0) < (m.awayGoal ?? 0)) awayWins++;
    else draws++;
  }
  const n = list.length;
  return {
    matches: n,
    totalGoals,
    averageGoalsPerMatch: n ? totalGoals / n : 0,
    homeWins,
    awayWins,
    draws,
    homeWinRate: n ? homeWins / n : 0,
    awayWinRate: n ? awayWins / n : 0,
    drawRate: n ? draws / n : 0,
  };
}

export function biggestWins(matches: Match[], filter: StatsFilter & { limit?: number } = {}): Match[] {
  const list = filterMatches(matches, filter);
  list.sort((a, b) => {
    const da = Math.abs((a.homeGoal ?? 0) - (a.awayGoal ?? 0));
    const db = Math.abs((b.homeGoal ?? 0) - (b.awayGoal ?? 0));
    if (db !== da) return db - da;
    const ga = (a.homeGoal ?? 0) + (a.awayGoal ?? 0);
    const gb = (b.homeGoal ?? 0) + (b.awayGoal ?? 0);
    return gb - ga;
  });
  return list.slice(0, filter.limit ?? 10);
}

export interface TeamSeasonScore {
  team: string;
  goalsFor: number;
  matches: number;
}

export function topScoringTeams(matches: Match[], filter: StatsFilter & { limit?: number } = {}): TeamSeasonScore[] {
  const list = filterMatches(matches, filter);
  const tally = new Map<string, TeamSeasonScore>();
  for (const m of list) {
    const home = tally.get(m.homeTeam) ?? { team: m.homeTeam, goalsFor: 0, matches: 0 };
    home.goalsFor += m.homeGoal ?? 0; home.matches++; tally.set(m.homeTeam, home);
    const away = tally.get(m.awayTeam) ?? { team: m.awayTeam, goalsFor: 0, matches: 0 };
    away.goalsFor += m.awayGoal ?? 0; away.matches++; tally.set(m.awayTeam, away);
  }
  const rows = Array.from(tally.values()).sort((a, b) => b.goalsFor - a.goalsFor);
  return rows.slice(0, filter.limit ?? 10);
}

export interface TeamVenueRecord {
  team: string;
  matches: number;
  wins: number;
  draws: number;
  losses: number;
  winRate: number;
}

export function bestRecord(
  matches: Match[],
  venue: "home" | "away",
  filter: StatsFilter & { limit?: number; minMatches?: number } = {},
): TeamVenueRecord[] {
  const list = filterMatches(matches, filter);
  const tally = new Map<string, TeamVenueRecord>();
  for (const m of list) {
    const teamName = venue === "home" ? m.homeTeam : m.awayTeam;
    const teamGoals = venue === "home" ? m.homeGoal ?? 0 : m.awayGoal ?? 0;
    const oppGoals = venue === "home" ? m.awayGoal ?? 0 : m.homeGoal ?? 0;
    const row = tally.get(teamName) ?? {
      team: teamName, matches: 0, wins: 0, draws: 0, losses: 0, winRate: 0,
    };
    row.matches++;
    if (teamGoals > oppGoals) row.wins++;
    else if (teamGoals < oppGoals) row.losses++;
    else row.draws++;
    tally.set(teamName, row);
  }
  const minMatches = filter.minMatches ?? 5;
  const rows = Array.from(tally.values())
    .filter((r) => r.matches >= minMatches)
    .map((r) => ({ ...r, winRate: r.wins / r.matches }))
    .sort((a, b) => b.winRate - a.winRate || b.wins - a.wins);
  return rows.slice(0, filter.limit ?? 10);
}

File: experiments/adrianco/experiment-5/runs/language=python_model=claude-opus-4-7_tooling=beads/rep2/src/soccer_mcp/stats.py
"""Aggregate statistics over the combined match table.

These are the "what's the average / which team is best at X" computations.
Each function returns plain dicts or lists so they can be ferried through the
MCP tool layer without further conversion. Pandas is used for the grouping
math; results are coerced back to Python types before returning.
"""

from __future__ import annotations

import pandas as pd

from soccer_mcp.data import SoccerData


def _scope(df: pd.DataFrame, competition: str | None, season: int | None) -> pd.DataFrame:
    if competition:
        df = df[df["competition"].str.lower().str.contains(competition.lower(), na=False)]
    if season is not None:
        df = df[df["season"] == season]
    return df.dropna(subset=["home_goal", "away_goal"])


def goals_per_match(data: SoccerData, competition: str | None = None, season: int | None = None) -> dict:
    df = _scope(data.matches, competition, season)
    if df.empty:
        return {"competition": competition, "season": season, "matches": 0, "average_goals": 0.0}
    avg = float((df["home_goal"] + df["away_goal"]).mean())
    return {
        "competition": competition,
        "season": season,
        "matches": int(len(df)),
        "average_goals": round(avg, 3),
        "home_goals_avg": round(float(df["home_goal"].mean()), 3),
        "away_goals_avg": round(float(df["away_goal"].mean()), 3),
    }


def home_advantage(data: SoccerData, competition: str | None = None, season: int | None = None) -> dict:
    """Home-win/draw/away-win rates and home win percentage in scope."""
    df = _scope(data.matches, competition, season)
    if df.empty:
        return {"matches": 0}
    home_wins = int((df["home_goal"] > df["away_goal"]).sum())
    away_wins = int((df["home_goal"] < df["away_goal"]).sum())
    draws = int((df["home_goal"] == df["away_goal"]).sum())
    total = home_wins + away_wins + draws
    return {
        "competition": competition,
        "season": season,
        "matches": total,
        "home_wins": home_wins,
        "away_wins": away_wins,
        "draws": draws,
        "home_win_rate": round(home_wins / total, 4) if total else 0.0,
        "away_win_rate": round(away_wins / total, 4) if total else 0.0,
        "draw_rate": round(draws / total, 4) if total else 0.0,
    }


def best_home_record(
    data: SoccerData,
    competition: str | None = None,
    season: int | None = None,
    min_matches: int = 5,
    limit: int = 10,
) -> list[dict]:
    """Teams with the best home win-rate (minimum match threshold)."""
    df = _scope(data.matches, competition, season)
    if df.empty:
        return []
    grouped = df.groupby("home_team_norm").apply(
        lambda g: pd.Series(
            {
                "team": g["home_team"].iloc[0],
                "matches": len(g),
                "wins": int((g["home_goal"] > g["away_goal"]).sum()),
                "draws": int((g["home_goal"] == g["away_goal"]).sum()),
                "losses": int((g["home_goal"] < g["away_goal"]).sum()),
            }
        ),
        include_groups=False,
    )
    grouped = grouped[grouped["matches"] >= min_matches]
    grouped["win_rate"] = grouped["wins"] / grouped["matches"]
    grouped = grouped.sort_values("win_rate", ascending=False).head(limit)
    return [
        {
            "team": row["team"],
            "matches": int(row["matches"]),
            "wins": int(row["wins"]),
            "draws": int(row["draws"]),
            "losses": int(row["losses"]),
            "win_rate": round(float(row["win_rate"]), 4),
        }
        for _, row in grouped.iterrows()
    ]


def best_away_record(
    data: SoccerData,
    competition: str | None = None,
    season: int | None = None,
    min_matches: int = 5,
    limit: int = 10,
) -> list[dict]:
    """Teams with the best away win-rate (minimum match threshold)."""
    df = _scope(data.matches, competition, season)
    if df.empty:
        return []
    grouped = df.groupby("away_team_norm").apply(
        lambda g: pd.Series(
            {
                "team": g["away_team"].iloc[0],
                "matches": len(g),
                "wins": int((g["away_goal"] > g["home_goal"]).sum()),
                "draws": int((g["away_goal"] == g["home_goal"]).sum()),
                "losses": int((g["away_goal"] < g["home_goal"]).sum()),
            }
        ),
        include_groups=False,
    )
    grouped = grouped[grouped["matches"] >= min_matches]
    grouped["win_rate"] = grouped["wins"] / grouped["matches"]
    grouped = grouped.sort_values("win_rate", ascending=False).head(limit)
    return [
        {
            "team": row["team"],
            "matches": int(row["matches"]),
            "wins": int(row["wins"]),
            "draws": int(row["draws"]),
            "losses": int(row["losses"]),
            "win_rate": round(float(row["win_rate"]), 4),
        }
        for _, row in grouped.iterrows()
    ]


def season_comparison(data: SoccerData, season_a: int, season_b: int, competition: str | None = None) -> dict:
    """Compare per-season match counts, goals, and home advantage."""
    return {
        "season_a": goals_per_match(data, competition=competition, season=season_a)
        | home_advantage(data, competition=competition, season=season_a),
        "season_b": goals_per_match(data, competition=competition, season=season_b)
        | home_advantage(data, competition=competition, season=season_b),
    }
